File: src-tauri/src/commands/snapshots.rs
use rusqlite::{params, OptionalExtension};
use tauri::State;

use crate::db::{new_id, now_ms, Db};
use crate::error::{Result, ScriptzError};
use crate::models::{Snapshot, SnapshotMeta};

const MAX_SNAPSHOTS_PER_SCRIPT: i64 = 50;

#[tauri::command]
pub fn create_snapshot(db: State<Db>, script_id: String, trigger: String) -> Result<SnapshotMeta> {
    let mut conn = db.conn()?;
    let trigger = if trigger == "manual" { "manual" } else { "auto" };
    let cj: String = conn
        .query_row(
            "SELECT content_json FROM scripts WHERE id = ?1",
            params![script_id],
            |r| r.get(0),
        )
        .optional()?
        .ok_or_else(|| ScriptzError::NotFound(format!("script {script_id}")))?;
    let id = new_id();
    let now = now_ms();
    let tx = conn.transaction()?;
    tx.execute(
        "INSERT INTO snapshots (id, script_id, content_json, trigger, created_at) VALUES (?1, ?2, ?3, ?4, ?5)",
        params![id, script_id, cj, trigger, now],
    )?;
    // Cap to MAX_SNAPSHOTS_PER_SCRIPT, oldest first
    tx.execute(
        "DELETE FROM snapshots WHERE id IN (
            SELECT id FROM snapshots WHERE script_id = ?1
            ORDER BY created_at DESC
            LIMIT -1 OFFSET ?2
         )",
        params![script_id, MAX_SNAPSHOTS_PER_SCRIPT],
    )?;
    tx.commit()?;
    Ok(SnapshotMeta {
        id,
        script_id,
        trigger: trigger.to_string(),
        created_at: now,
    })
}

#[tauri::command]
pub fn list_snapshots(db: State<Db>, script_id: String) -> Result<Vec<SnapshotMeta>> {
    let conn = db.conn()?;
    let mut stmt = conn.prepare(
        "SELECT id, script_id, trigger, created_at FROM snapshots WHERE script_id = ?1 ORDER BY created_at DESC",
    )?;
    let rows = stmt.query_map(params![script_id], |r| {
        Ok(SnapshotMeta {
            id: r.get(0)?,
            script_id: r.get(1)?,
            trigger: r.get(2)?,
            created_at: r.get(3)?,
        })
    })?;
    let mut out = Vec::new();
    for r in rows {
        out.push(r?);
    }
    Ok(out)
}

#[tauri::command]
pub fn get_snapshot(db: State<Db>, id: String) -> Result<Snapshot> {
    let conn = db.conn()?;
    conn.query_row(
        "SELECT id, script_id, content_json, trigger, created_at FROM snapshots WHERE id = ?1",
        params![id],
        |r| {
            Ok(Snapshot {
                id: r.get(0)?,
                script_id: r.get(1)?,
                content_json: r.get(2)?,
                trigger: r.get(3)?,
                created_at: r.get(4)?,
            })
        },
    )
    .optional()?
    .ok_or_else(|| ScriptzError::NotFound(format!("snapshot {id}")))
}

#[tauri::command]
pub fn restore_snapshot(db: State<Db>, snapshot_id: String) -> Result<()> {
    let mut conn = db.conn()?;
    let now = now_ms();
    let (script_id, content_json): (String, String) = conn
        .query_row(
            "SELECT script_id, content_json FROM snapshots WHERE id = ?1",
            params![snapshot_id],
            |r| Ok((r.get::<_, String>(0)?, r.get::<_, String>(1)?)),
        )
        .optional()?
        .ok_or_else(|| ScriptzError::NotFound(format!("snapshot {snapshot_id}")))?;

    // Backup-snapshot + content overwrite + FTS refresh in one
    // transaction so we never end up with an orphan backup snapshot
    // pointing at content that wasn't actually replaced.
    let tx = conn.transaction()?;
    let cur_cj: String = tx.query_row(
        "SELECT content_json FROM scripts WHERE id = ?1",
        params![script_id],
        |r| r.get(0),
    )?;
    let backup_id = new_id();
    tx.execute(
        "INSERT INTO snapshots (id, script_id, content_json, trigger, created_at) VALUES (?1, ?2, ?3, 'auto', ?4)",
        params![backup_id, script_id, cur_cj, now],
    )?;
    tx.execute(
        "UPDATE scripts SET content_json = ?1, updated_at = ?2 WHERE id = ?3",
        params![content_json, now, script_id],
    )?;
    crate::commands::scripts::refresh_fts_for_script(&tx, &script_id)?;
    tx.commit()?;
    Ok(())
}

#[tauri::command]
pub fn delete_snapshot(db: State<Db>, id: String) -> Result<()> {
    let conn = db.conn()?;
    conn.execute("DELETE FROM snapshots WHERE id = ?1", params![id])?;
    Ok(())
}

File: apps/desktop/src/lib/snapshots.ts
// Snapshots — auto + manual versioning for a script's content.
//
// TS port of src-tauri/src/commands/snapshots.rs (Migration Phase 5).
// Schema is unchanged: `snapshots(id, script_id, content_json, trigger,
// created_at)`, FK to `scripts.id` ON DELETE CASCADE. Cap is 50 per
// script (oldest dropped automatically on each create).
//
// Caveat (same as earlier phases): tauri-plugin-sql has no transaction
// API in JS, so multi-statement operations run as independent
// auto-committed steps. Effects:
//   - createSnapshot: a crash between INSERT and the trim DELETE
//     could leave a 51st row briefly; the next create_snapshot trims
//     it. No data loss either way.
//   - restoreSnapshot: backup INSERT + content UPDATE + FTS refresh
//     are sequential. A crash mid-flow leaves the script either at
//     its old content (if before UPDATE) or at the new content with
//     a fresh backup snapshot present (if after). Worst case the FTS
//     index lags one save behind, which the next save corrects.

import { getDb } from "./db";
import { refreshFtsForScript } from "./fts";
import type { Snapshot, SnapshotMeta } from "./types";

const MAX_SNAPSHOTS_PER_SCRIPT = 50;

export async function createSnapshot(
  scriptId: string,
  trigger: "auto" | "manual",
): Promise<SnapshotMeta> {
  const db = await getDb();
  const triggerNorm: "auto" | "manual" = trigger === "manual" ? "manual" : "auto";

  const rows = await db.select<{ content_json: string }[]>(
    "SELECT content_json FROM scripts WHERE id = $1",
    [scriptId],
  );
  if (rows.length === 0) {
    throw new Error(`not found: script ${scriptId}`);
  }
  const contentJson = rows[0].content_json;

  const id = crypto.randomUUID();
  const now = Date.now();
  await db.execute(
    "INSERT INTO snapshots (id, script_id, content_json, trigger, created_at) VALUES ($1, $2, $3, $4, $5)",
    [id, scriptId, contentJson, triggerNorm, now],
  );
  // Cap to MAX_SNAPSHOTS_PER_SCRIPT, oldest first.
  await db.execute(
    `DELETE FROM snapshots WHERE id IN (
       SELECT id FROM snapshots WHERE script_id = $1
       ORDER BY created_at DESC
       LIMIT -1 OFFSET $2
     )`,
    [scriptId, MAX_SNAPSHOTS_PER_SCRIPT],
  );

  return {
    id,
    script_id: scriptId,
    trigger: triggerNorm,
    created_at: now,
  };
}

export async function listSnapshots(scriptId: string): Promise<SnapshotMeta[]> {
  const db = await getDb();
  return db.select<SnapshotMeta[]>(
    `SELECT id, script_id, trigger, created_at FROM snapshots
     WHERE script_id = $1 ORDER BY created_at DESC`,
    [scriptId],
  );
}

export async function getSnapshot(id: string): Promise<Snapshot> {
  const db = await getDb();
  const rows = await db.select<Snapshot[]>(
    "SELECT id, script_id, content_json, trigger, created_at FROM snapshots WHERE id = $1",
    [id],
  );
  if (rows.length === 0) {
    throw new Error(`not found: snapshot ${id}`);
  }
  return rows[0];
}

export async function restoreSnapshot(snapshotId: string): Promise<void> {
  const db = await getDb();
  const rows = await db.select<{ script_id: string; content_json: string }[]>(
    "SELECT script_id, content_json FROM snapshots WHERE id = $1",
    [snapshotId],
  );
  if (rows.length === 0) {
    throw new Error(`not found: snapshot ${snapshotId}`);
  }
  const { script_id: scriptId, content_json: restoredContent } = rows[0];

  const now = Date.now();

  // Take an auto-snapshot of the current content before overwriting.
  const curRows = await db.select<{ content_json: string }[]>(
    "SELECT content_json FROM scripts WHERE id = $1",
    [scriptId],
  );
  if (curRows.length === 0) {
    throw new Error(`not found: script ${scriptId}`);
  }
  const backupId = crypto.randomUUID();
  await db.execute(
    "INSERT INTO snapshots (id, script_id, content_json, trigger, created_at) VALUES ($1, $2, $3, 'auto', $4)",
    [backupId, scriptId, curRows[0].content_json, now],
  );
  // Restore counts as a snapshot insert; honour the 50-per-script cap so
  // repeated restores (or restore-after-many-saves) can't grow the table
  // unbounded. Mirrors createSnapshot's trim DELETE.
  await db.execute(
    `DELETE FROM snapshots WHERE id IN (
       SELECT id FROM snapshots WHERE script_id = $1
       ORDER BY created_at DESC
       LIMIT -1 OFFSET $2
     )`,
    [scriptId, MAX_SNAPSHOTS_PER_SCRIPT],
  );

  await db.execute(
    "UPDATE scripts SET content_json = $1, updated_at = $2 WHERE id = $3",
    [restoredContent, now, scriptId],
  );
  await refreshFtsForScript(scriptId);
}

export async function deleteSnapshot(id: string): Promise<void> {
  const db = await getDb();
  await db.execute("DELETE FROM snapshots WHERE id = $1", [id]);
}
